### PR TITLE
DAOS-5877 rebuild: Only merge queued rebuild tasks for sequential operations

### PR DESCRIPTION
**Pull request by byron-marohn**
~~This is important for correct handling when multiple system changes
happen very close together.~~

~~Also adds a nice rebuild queue print function so that the queue can be
examined if debug logs are turned on to help diagnose problems.~~

Prior to this patch, rebuild will merge any tasks together that are of
the same operation. This used to always be a safe thing to do when
rebuild handled only failure recovery. However, now that rebuild handles
multiple different types of operations, it was still able to merge the
same type together, even if there was a separate operation that happened
in between. This caused the rebuild version to be set incorrectly and
placement to return incorrect data, resulting in data loss.

With this fix, rebuild will only merge tasks that are safe to merge
because the same type of operation happened after another of the same
type without any other types in between. For example, fail -> fail ->
fail will merge into just one task, but fail -> reint -> fail will leave
three independent queued tasks, which are executed in that order.
Without this patch, this would merge the two fails together and handle
them both prior to the reintegration, which was incorrect.

### CODE REVIEW

**Review by ryon-jensen — DISMISSED** [review]
LGTM

File: src/rebuild/srv.c [hunk 1]
```diff
@@ -1378,6 +1383,103 @@ rebuild_print_list_update(const char *const str, const uuid_t uuid,
...
 }
 
+/* Here we want to schedule this new rebuild task by placing it on the
+ * queue. However, the ordering is important - tasks will be consumed
+ * by iterating the queue and running the first available one for each
+ * pool. So what we do here is iterate through the queue, and insert
+ * this task before the first other found task for the same pool that
+ * should run with lower priority. By doing this, the list will always
+ * be sorted
+ */
```
Comment (NiuYawei) [inline src/rebuild/srv.c:1393]: Can we do these rebuild, drain, reint, extend operations despite of f_seq / pool version order?

Comment (byron-marohn) [reply src/rebuild/srv.c]: This was an excellent catch, and indeed we can not do this - and it uncovered the actual bug here (which is that rebuild was allowed to merge tasks together in such a way as to make the versions incorrect. Thanks a bunch for catching this!

**Comment by byron-marohn** [comment]
Pushed a completely new fix for this, changed original message & the title of this pull request. It now does not do any prioritization - and it also fixes the bug that Niu discovered where rebuild was combining things that could not be safely combined.

**Review by NiuYawei — APPROVED** [review]

File: src/rebuild/srv.c [hunk 2]
```diff
@@ -956,8 +1029,12 @@ rebuild_try_merge_tgts(const uuid_t pool_uuid, uint32_t map_ver,
 		found->dst_map_ver = map_ver;
```
Comment (gnailzenh) [inline src/rebuild/srv.c:1029]: Not from this patch, but in the "if" statement, the variable is "task", but this line is using "found", I think they are the same, but it's still better to use the same variable?

File: src/rebuild/srv.c [hunk 3]
```diff
@@ -926,13 +973,39 @@ rebuild_try_merge_tgts(const uuid_t pool_uuid, uint32_t map_ver,
...
+	 * safe to merge to, even if it's the same type, because this will cause
+	 * rebuild to essentially skip the intermediary different-type step
+	 * because the rebuild version is set to the task map version after
+	 * rebuild is complete.
+	 */
+	d_list_for_each_entry(task, &rebuild_gst.rg_queue_list, dst_list) {
 		if (uuid_compare(task->dst_pool_uuid, pool_uuid) == 0
-		    && task->dst_rebuild_op == rebuild_op) {
+		    && task->dst_rebuild_op == rebuild_op
+		    && task->dst_map_ver == highest_queued_map_ver) {
```
Comment (gnailzenh) [inline src/rebuild/srv.c:1008]: you could do this in one loop by change the "if" 
```
do_list_for_each_entry(..) {
    if (uuid_compare(...) != 0)
           continue;
     if (task->dst_rebuild_op != rebuild_op)
            found = NULL;
     else
            found = task;
 }
```
Now you don't need to previous loop correct?

**Review by gnailzenh — DISMISSED** [review]
Just a few comments

File: src/rebuild/srv.c [hunk 4]
```diff
@@ -913,7 +913,54 @@ rebuild_task_destroy(struct rebuild_task *task)
...
+/**
+ * Print out all of the currently queued rebuild tasks
+ */
+static void
+rebuild_debug_print_queue()
+{
+	struct rebuild_task *task_tmp;
+	struct rebuild_task *task;
+	/* Uninitialized stack buffer to write target list into */
+	char tgts_buf[512];
```
[the remark this answers is not in the document]
Comment (byron-marohn) [reply src/rebuild/srv.c]: Reduced the size to 200 bytes and added a comment. Still plenty of space for >30 5-digit target IDs. Allocation just makes this more complicated than it already is for little value.

File: src/rebuild/srv.c [hunk 5]
```diff
@@ -913,7 +913,54 @@ rebuild_task_destroy(struct rebuild_task *task)
...
+	/* Position in stack buffer where str data should be written next */
+	size_t tgts_pos = 0;
+
+	D_DEBUG(DB_REBUILD, "Current rebuild queue:\n");
+
+	d_list_for_each_entry_safe(task, task_tmp,
+				   &rebuild_gst.rg_queue_list, dst_list) {
+
+		for (i = 0; i < task->dst_tgts.pti_number; i++) {
+			if (tgts_pos > sizeof(tgts_buf) - 10) {
```
[the remark this answers is not in the document]
Comment (byron-marohn) [reply src/rebuild/srv.c:939]: Added a comment for this too. Just wanted to make sure if you were going to print a 6-digit target ID that there'd still be space for the ... to get appended.

File: src/rebuild/srv.c (hunk 3 above)
Comment (byron-marohn) [reply src/rebuild/srv.c]: Good observation. I've refactored this as you suggest, which should work.

File: src/rebuild/srv.c (hunk 2 above)
Comment (byron-marohn) [reply src/rebuild/srv.c]: task and found were only the same with the "break" statement. I renamed found to "merge_task" and replaced references to "task" with "merge_task".

**Review by NiuYawei — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]

**Comment by byron-marohn** [comment]
Test failures are these: 
```
Test / Functional on Leap 15 / 1-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_objects;large-hosts-pool-rebuild-servers-950e – RebuildDeleteObjects
1m 33s
Test / Functional on Leap 15 / 2-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_objects;small-hosts-pool-rebuild-servers-06b4 – RebuildDeleteObjects
1m 6s
Test / Functional on Leap 15 / 3-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_records;large-hosts-pool-rebuild-servers-950e – RebuildDeleteObjects
1m 37s
Test / Functional on Leap 15 / 4-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_records;small-hosts-pool-rebuild-servers-06b4 – RebuildDeleteObjects
1m 3s
Test / Functional on CentOS 7 / 1-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_objects;large-hosts-pool-rebuild-servers-fd0c – RebuildDeleteObjects
1m 29s
Test / Functional on CentOS 7 / 2-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_objects;small-hosts-pool-rebuild-servers-0f1c – RebuildDeleteObjects
1m 1s
Test / Functional on CentOS 7 / 3-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_records;large-hosts-pool-rebuild-servers-fd0c – RebuildDeleteObjects
1m 32s
Test / Functional on CentOS 7 / 4-./rebuild/delete_objects.py:RebuildDeleteObjects.test_rebuild_delete_records;small-hosts-pool-rebuild-servers-0f1c – RebuildDeleteObjects
1m 1s```

Which are all tracked by DAOS-6118, and not related to this PR.

**Comment by byron-marohn** [comment]
Created DAOS-6429, which tracks the segfault seen in this recent run. Was also hit in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4175/3/tests with these same segfault failures.

**Comment by byron-marohn** [comment]
@daos-stack/daos-gatekeeper this is ready to land - passed CI and two approvals.

File: src/rebuild/srv.c [hunk 6]
```diff
@@ -922,7 +922,61 @@ rebuild_task_destroy(struct rebuild_task *task)
 	D_FREE(task);
 }
 
+/**
+ * Print out all of the currently queued rebuild tasks
+ */
+static void
+rebuild_debug_print_queue()
```
Comment (ashleypittman) [inline src/rebuild/srv.c:929]: For what it's worth, this is similar to the d_rank_list_dump() function.

Comment (ashleypittman) [reply src/rebuild/srv.c:929]: The types are different but rank lists are arrays, not linked lists.

File: src/rebuild/srv.c [hunk 7]
```diff
@@ -922,7 +922,61 @@ rebuild_task_destroy(struct rebuild_task *task)
...
+{
+	struct rebuild_task *task_tmp;
+	struct rebuild_task *task;
+	/* Uninitialized stack buffer to write target list into
+	 * This only accumulates the targets in a single task, so it doesn't
+	 * need to be very big. 200 bytes is enough for ~30 5-digit target ids
+	 */
+	char tgts_buf[200];
```
Comment (gnailzenh) [inline src/rebuild/srv.c:937]: I still think it's better to use a __thread static variable for this, but no big deal and can be addressed later.

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**